### PR TITLE
task 1&2 (test pending)

### PR DESCRIPTION
**Pull request by handmilkingsoftware**
Angular app has a sandbox in order to fake twilio messages from a phone number.

### CODE REVIEW

File: vreasy/application/modules/vreasy/controllers/TwilioController.php [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+    private function guessAnswer($message)
+    {
+        $reject_pattern = "/wouldn't be able|would not be able|can't|cannot|cant|can not|no|do not|n�o|non|dim|nein/i";
+        $accept_pattern = "/great|sure|why not|right away|I'm here to obey|okey makey|absolutely|like a shot|at once|now|straightaway|directly|straight off|forthwith|instantly|immediately|promptly|ja|si|yes|oui|s�/i";
+        $finish_pattern = "/finish|done|completado|completed|finished|mission accomplished|termin�|have done,ve done|work done|fet|fertig|termin�|feito/i";
+        $working_pattern = "/working|functioning|operation|operating|running|action|performing|machinery|realiza��o|Durchf�hrung|de travail/i";
+        $result = array();
+
+        if (preg_match($accept_pattern, $message)) {
```
Comment (bbwebconsult) [inline vreasy/application/modules/vreasy/controllers/TwilioController.php]: @handmilkingsoftware 
1) What would happen if the user misspells his answer. e.g.: `shure` (instead of sure)
2) What do you think the result would be if the user responds `impossible`

How would you handle this?

File: library/Vreasy/Models/Task.php [hunk 2]
```diff
@@ -39,50 +143,112 @@ public function save()
...
+                'from' => array('working', 'accepted'),
+                'to' => NULL
+            )
+        );
+        $ends = $statusFlow->$status;
+        return ($end === 'prev') ? $ends->from : $ends->to;
     }
 
+    //U
+    public function updateStatus($status, $msg, $manual = FALSE)
```
Comment (bbwebconsult) [inline library/Vreasy/Models/Task.php]: @handmilkingsoftware why a separate method for saving the status?
Also, I did not quite understand that "1/4, 1/2 and 3/4" rule that I thinkyou try to explain [here](https://github.com/Vreasy/VrRecruit/pull/20/files#diff-04c6e90faac2675aa89e2176d2eec7d8R16)
Could you develop?

File: vreasy/application/modules/vreasy/controllers/TwilioController.php (hunk 1 above)
Comment (handmilkingsoftware) [inline vreasy/application/modules/vreasy/controllers/TwilioController.php]: 1)
Sure.
At the moment the patterns doens't accept words our of this subset of words....

At the moment the webapp would proceed like this:
1. mark it as "undefined"
2. send a feedback to the user letting him know the last message was not valid (see TwilioController+Line#186). This kind of feedback also happens if the provider tries to accept a "non issued" task, or mark as completed a "non accepted task"
3. it would get logged into the log file.

The path to follow would be:
A. Statistically and manually select the error and typos that have been collected since the app would be available and store then to select which would match as valid and add them to the regular expression pattern.
B. To get a most robust misspelling handling I would add typo tollerance through levenshtein() (similar_text() or any other solution) but that only if the word doesn't fit into a pre-selected word sub-set. Reason is these kind of functions has a considerable cost which should be avoided.
http://us2.php.net/manual/en/function.levenshtein.php

Comment (handmilkingsoftware) [inline vreasy/application/modules/vreasy/controllers/TwilioController.php]: 2)
It would fail matching a valid response... Would get a feedback and log it into logs.

File: library/Vreasy/Models/Task.php (hunk 2 above)
Comment (handmilkingsoftware) [inline library/Vreasy/Models/Task.php]: ok when task get feed into the database from the migration script.
1) That task get a DUE date randomly selected from anywhere withing the next 15 days.
2) The issue status (log status) is set to current time.
3. The migration script also calls Tasks (model) method insertMissingLogs. So if the task being inserted from the migration is "completed": insertMissingLogs() woudl insert the issue and accepted mising history (logs) states. Just to fill what is missing.

Under this scenario, if a task entry was "completed" for instance at migrations/webapp/20150811100000_InsertSomeTasks.php::LINE 14
(object)array("ico" => "wrench", "description" => "Set up Andre's working place.", 'deadline' => (new \DateTime(getRandomDate() . " seconds"))->format(DATE_FORMAT), 'status' => 'completed', "idmember" => 1),

The due (deadline) date at this point is unknown while the issue time (current time) is known
in order to fill issue and accepted states (and lately completed) in the log I came out with this "quatter, half and 3quatter" of time
1 quatter = (duetime-currenttime)/4
1 half = (duetime-currenttime)/2
and so on...

this let me put issue time at current time
this let me put accepted time at 1 quarter
this let me put working time at 1 half
this let me put completed time at 3 quarters.

-----------
So if i come out and put completed status (manual or through twilio msg) the current time would be ignore (if not completed status would be next to issue instead greater than working)
so if you had a "working task" and you come to put it in completed, it would override current time to maintain coherence in the timeline

conclusion
All completed status are a 3 quarter
All working task status get stablished exactly in the middle from when the task got issued and the deadline date.
All rejects and accept status get stablished exactly one quarter of the time from a distance between the task got issued and the deadline.

You can access the live version at
http://201.237.63.80 don't know if picard let you know that ip address before.

Comment (handmilkingsoftware) [inline library/Vreasy/Models/Task.php]: why a separate method for saving the status?
Reason is it gets a validation before proceed
There is some diagrams called "state digrams" (UML).
Aplied to this scenario, all status has a "FROM" status and a "TO" status, or just previous state and next state.
This way you cannot move a state from "accepted" to "issued" for instance (you would need helping state like cancel).
The same way you wouldn't be able to move from completed to other states.
This constrains are enforced through a method called "statusFlow" at line 162 withing Task model.
That very method is called by "updateStatus" method which is used for save method(conditionally) but also for "insertMissingLogs()" which as pointed before is called from the "migration scripts".

So. To change the status of a task it first has a state flow validation and later not only modified the "tasks" table but also the "task_history" table.

Comment (handmilkingsoftware) [inline library/Vreasy/Models/Task.php]: what did you mean with "Could you develop?". Sort of typo or something? thanks for the time reviewing the code.

File: vreasy/application/configs/db.ini [hunk 3]
```diff
@@ -10,8 +10,8 @@ database.params.charset = "utf8"
 database.adapter = "PDO_MYSQL"
 database.params.dbname = vreasy_task_confirmation
 database.params.host = 127.0.0.1
-database.params.username = "vreasy"
-database.params.password = ";FeA336101-vreasy_task_confirmation"
+database.params.username = "root"
+database.params.password = ""
 database.params.charset = "utf8"
```
Comment (gpou) [inline vreasy/application/configs/db.ini]: Hello Andres,
I'm taking the time to fetch your PR in order to test it locally.
Why did you change the database user and password?

**Comment by bbwebconsult** [comment]
Thank you @handmilkingsoftware for taking the time to do the test, We'll come back to you soon.

File: vreasy/application/configs/db.ini (hunk 3 above)
Comment (gpou) [inline vreasy/application/configs/db.ini]: @handmilkingsoftware You shouldn't rename the already existing migrations. Instead, write new migrations to change/add/remove the fields as needed.

Logically, the migration AddTasksTable throws an error Table 'tasks' already exists

Comment (mwpicard) [inline vreasy/application/configs/db.ini]: He's in Costa Rica (and cant relocate). Not Awesome PLUS remote = do not
hire

correct?

Or can u think of someting easy to delegate?

File: vreasy/application/modules/vreasy/controllers/TwilioController.php (hunk 1 above)
Comment (bbwebconsult) [inline vreasy/application/modules/vreasy/controllers/TwilioController.php]: @handmilkingsoftware I meant to say (in a kind of hidden way :smile: ): Wouldn't "impos**si**ble" match `si` in `$accept_pattern` and thus be considered as a positive value?

File: migrations/webapp/20150810100000_AddTasksTable.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            ]
+        );
+        $tasks->column('idmember', 'integer', [
+            'unsigned' => TRUE,
+            'null' => FALSE
+        ]);
+
+        $tasks->finish();
+        $this->execute("ALTER TABLE `vreasy_task_confirmation`.`tasks` ADD INDEX `fk_tasks_staff_idx` (`idmember` ASC); ALTER TABLE `vreasy_task_confirmation`.`tasks` ADD CONSTRAINT `fk_tasks_staff` FOREIGN KEY (`idmember`) REFERENCES `vreasy_task_confirmation`.`staff` (`memberid`) ON DELETE NO ACTION ON UPDATE NO ACTION;");
+
```
Comment (gpou) [inline migrations/webapp/20150810100000_AddTasksTable.php]: @handmilkingsoftware Inserting the databasename here makes the migration to fail on test env.
You can use this:
        $this->add_foreign_key(
            'tasks',
            'staff', [
            'orig_column_name' => 'idmember',
            'dest_column_name' => 'memberid',
            'on_delete' => 'NO ACTION',
            'on_update' => 'NO ACTION',
        ]);


I guess we are using a branch of ruckusing:
https://github.com/ruckus/ruckusing-migrations/pull/157/files#diff-ddabc755cd12c0955d69bb6f9eedcd0eR892

**Comment by gpou** [comment]
@handmilkingsoftware Once everything installed, I get a lot of errors "Error in resource configuration for action `query`. Expected response to contain an array but got an object"

Maybe a problem of versions?

Please  provide detailed instructions about installation of all the dependencies.

File: migrations/webapp/20150810100000_AddTasksTable.php (hunk 4 above)
Comment (handmilkingsoftware) [inline migrations/webapp/20150810100000_AddTasksTable.php]: haha I actually didn't know how properly insert FKs in ruckusing so I did plain execute MySQL sentence to link the Foreign key which in my opinion is equivalent. But you are absolutely right, I should had omited the database Name.

File: vreasy/application/configs/db.ini (hunk 3 above)
Comment (handmilkingsoftware) [inline vreasy/application/configs/db.ini]: I know but since the actual table (tasks) got broken into several pieces, for instance phone number and provider name got reallocated into "staff" table I found it kind of ungly drop those columns and move take further actions to move them into staff table.
Of curse I understand this shouldn't happend in production enviroment but for this demo/test purposes I just found it "harmless"

**Comment by handmilkingsoftware** [comment]
Thanks for trying to run the project.
Let me check it further. First I tried to move all "ACTION" controllers to "ZEND_REST" controllers. Reason was I wasn't rendering anything on them but later came back to the actual controllers schema and tried to stick to the actual challenge and work within the ecosystem of vreasy's libraries you got there.
I'll fresh-clone the project and see what problems it brings out. Here I'll post my virtual host and htaccess and see if it helps

Another possible cause is that it tries to write to logs/logs.txt with a (append) mode and perhaps it fails to write if not yet created (or no right to write) I'll check all this details.

Also removed kind of innecesary (for this test) details like ?format=json query parameter.
Also moved index.php to api/index.php but most of these things doesn't seems to be the problem here. Going to check and come back to you asap.

**Comment by handmilkingsoftware** [comment]
My virtual host conf...
<VirtualHost *:80>
         DocumentRoot "E:/projects/vrRecruit/vreasy/public"
		 ErrorDocument 404 /index.html
         ServerName vreasydev.com
         SetEnv APPLICATION_ENV "development"
   <Directory "E:/projects/vrRecruit/vreasy/public">
         Options Indexes FollowSymLinks MultiViews
         AllowOverride all
         Order allow,deny
         Allow from all
    </Directory>
</VirtualHost>

**Comment by gpou** [comment]
@handmilkingsoftware I finally got it to work by adding a /vreasy/application/layouts/scripts/vreasy.phtml

Looks great!

I did some tests of sending a message with "yes" to an issued task, and got the task assigned, and under the history a new "accepted" log appeared (had to refresh the page, though)

I also tried to send a "cannot" message to an already accepted task, and got the corresponding "There is no issued tasks" message reply. Great.
I had to solve some little issues (https://github.com/Vreasy/VrRecruit/pull/20/files#diff-b2ae17568af2b2322d28ff3d076e69f7R190 throws an error of Undefined property: stdClass::$noTask in /home/gemma/vreasy/VrRecruit/library/Vreasy/Models/Task.php on line 186)

The dates for new history logs are all Tomorrow, but I haven't investigated the cause.

I think the goal of the test was to work with the twilio services, which is working perfectly, and to refactor the models to add the members, history, etc, which is also working.

You also implemented the statusFlow, which is a good way to ensure that the tasks' status follow a well defined way.

Good job Andrés!

**Comment by handmilkingsoftware** [comment]
please add vreasy/public/api empty folder that would solve it also...
since directory structure changes a bit

I replaced index.php for index.html and also index.php got moved to api/index.php..
